File: acrarium/src/test/java/com/faendir/acra/rest/RestReportInterfaceTest.java
/*
 * (C) Copyright 2018 Lukas Morawietz (https://github.com/F43nd1r)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.faendir.acra.rest;

import com.faendir.acra.model.App;
import com.faendir.acra.security.VaadinSessionSecurityContextHolderStrategy;
import com.faendir.acra.service.DataService;
import com.faendir.acra.service.UserService;
import junit.framework.AssertionFailedError;
import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.FilterType;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;

import java.util.Arrays;
import java.util.Optional;

import static com.faendir.acra.rest.RestReportInterface.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
import static org.springframework.http.MediaType.APPLICATION_OCTET_STREAM_VALUE;
import static org.springframework.http.MediaType.MULTIPART_FORM_DATA;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.multipart;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

/**
 * @author lukas
 * @since 26.06.18
 */
@RunWith(SpringRunner.class)
@WebMvcTest(controllers = RestReportInterface.class, includeFilters = @ComponentScan.Filter(classes = EnableWebSecurity.class), excludeFilters = @ComponentScan.Filter(type = FilterType.ASSIGNABLE_TYPE, value = VaadinSessionSecurityContextHolderStrategy.class))
@WithMockUser(roles = {"REPORTER", "USER"})
public class RestReportInterfaceTest {
    private final String TEST_STRING = "TEST";
    @MockBean UserService userService;
    @MockBean DataService dataService;
    @Autowired RestReportInterface restReportInterface;
    @Autowired MockMvc mvc;

    @Before
    public void setUp() {
        App app = new App();
        when(dataService.findApp(TEST_STRING)).thenReturn(app);
        when(dataService.getFromReports(any(), any(), any(), any())).thenReturn(Arrays.asList("{\"name\":\"a\"}", "{\"name\":\"b\"}"));
    }

    @Test
    public void report() throws Exception {
        mvc.perform(post("/" + REPORT_PATH).contentType(APPLICATION_JSON).content(TEST_STRING)).andExpect(status().isOk());
        verify(dataService, times(1)).createNewReport(any(), any(), any());
    }

    @Test
    public void report2() throws Exception {
        mvc.perform(multipart("/" + REPORT_PATH).file(new MockMultipartFile(REPORT, TEST_STRING, APPLICATION_JSON_VALUE, new byte[0]))
                .file(new MockMultipartFile(ATTACHMENT, TEST_STRING, APPLICATION_OCTET_STREAM_VALUE, new byte[0]))
                .contentType(MULTIPART_FORM_DATA)).andExpect(status().isOk());
        verify(dataService, times(1)).createNewReport(any(), any(), any());
    }

    @Test
    public void exportWithId() throws Exception {
        mvc.perform(get("/" + EXPORT_PATH).param(PARAM_APP, TEST_STRING).param(PARAM_ID, TEST_STRING)).andExpect(status().isOk()).andExpect(m -> {
            JSONArray array = new JSONArray(m.getResponse().getContentAsString());
            if (array.length() != 2 || !(array.get(0) instanceof JSONObject)) {
                throw new AssertionFailedError();
            }
        });
    }

    @Test
    public void exportWithMail() throws Exception {
        mvc.perform(get("/" + EXPORT_PATH).param(PARAM_APP, TEST_STRING).param(PARAM_MAIL, TEST_STRING)).andExpect(status().isOk()).andExpect(m -> {
            JSONArray array = new JSONArray(m.getResponse().getContentAsString());
            if (array.length() != 2 || !(array.get(0) instanceof JSONObject)) {
                throw new AssertionFailedError();
            }
        });
    }

    @Test
    public void exportInvalid() throws Exception {
        mvc.perform(get("/" + EXPORT_PATH).param(PARAM_APP, TEST_STRING)).andExpect(status().is4xxClientError());
        mvc.perform(get("/" + EXPORT_PATH).param(PARAM_ID, TEST_STRING)).andExpect(status().is4xxClientError());
        mvc.perform(get("/" + EXPORT_PATH).param(PARAM_MAIL, TEST_STRING)).andExpect(status().is4xxClientError());
        mvc.perform(get("/" + EXPORT_PATH)).andExpect(status().is4xxClientError());
    }
}

File: acrarium/src/test/kotlin/com/faendir/acra/rest/RestReportInterfaceTest.kt
/*
 * (C) Copyright 2018-2024 Lukas Morawietz (https://github.com/F43nd1r)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.faendir.acra.rest

import com.faendir.acra.domain.ReportService
import com.faendir.acra.persistence.app.AppId
import com.faendir.acra.persistence.bug.BugId
import com.faendir.acra.persistence.report.Report
import com.faendir.acra.persistence.user.UserRepository
import com.faendir.acra.rest.RestReportInterface.Companion.ATTACHMENT
import com.faendir.acra.rest.RestReportInterface.Companion.REPORT
import com.faendir.acra.rest.RestReportInterface.Companion.REPORT_PATH
import com.ninjasquad.springmockk.MockkBean
import com.vaadin.flow.spring.SpringServlet
import com.vaadin.flow.spring.security.VaadinDefaultRequestCache
import io.mockk.every
import io.mockk.verify
import org.jooq.JSON
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.boot.web.servlet.ServletRegistrationBean
import org.springframework.context.annotation.ComponentScan
import org.springframework.http.MediaType.*
import org.springframework.mock.web.MockMultipartFile
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.multipart
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.time.Instant

@WebMvcTest(controllers = [RestReportInterface::class], includeFilters = [ComponentScan.Filter(classes = [EnableWebSecurity::class])])
@WithMockUser(roles = ["REPORTER", "USER"])
class RestReportInterfaceTest {

    @MockkBean
    lateinit var requestCache: VaadinDefaultRequestCache

    @MockkBean
    lateinit var springServletServletRegistrationBean: ServletRegistrationBean<SpringServlet>

    @MockkBean
    lateinit var userRepository: UserRepository

    @MockkBean
    lateinit var reportService: ReportService

    @Autowired
    lateinit var restReportInterface: RestReportInterface

    @Autowired
    lateinit var mvc: MockMvc

    @BeforeEach
    fun setUp() {
        every { reportService.create(any(), any(), any()) } returns Report(
            id = "",
            androidVersion = null,
            content = JSON.json("{}"),
            date = Instant.now(),
            phoneModel = null,
            userComment = null,
            userEmail = null,
            brand = null,
            installationId = "",
            isSilent = false,
            device = "",
            marketingDevice = "",
            bugId = BugId(0),
            appId = AppId(0),
            stacktrace = "",
            exceptionClass = "",
            message = null,
            crashLine = null,
            cause = null,
            versionCode = 0,
            versionFlavor = ""
        )
        every { requestCache.getMatchingRequest(any(), any()) } returns null
    }

    @Test
    fun `should accept report without attachments`() {
        mvc.perform(post("/$REPORT_PATH").contentType(APPLICATION_JSON).content(TEST_STRING)).andExpect(status().isOk)
        verify(exactly = 1) { reportService.create(any(), any(), emptyList()) }
    }

    @Test
    fun `should accept report with attachments`() {
        mvc.perform(
            multipart("/$REPORT_PATH").file(MockMultipartFile(REPORT, TEST_STRING, APPLICATION_JSON_VALUE, ByteArray(0)))
                .file(MockMultipartFile(ATTACHMENT, TEST_STRING, APPLICATION_OCTET_STREAM_VALUE, ByteArray(0)))
                .contentType(MULTIPART_FORM_DATA)
        ).andExpect(status().isOk)
        verify(exactly = 1) { reportService.create(any(), any(), any()) }
    }

    companion object {
        private const val TEST_STRING = "TEST"
    }
}